shotover/src/codec/cassandra.rs

```diff
@@ -561,46 +657,108 @@ impl CassandraEncoder {
 
-                // add payload crc
-                dst.extend_from_slice(
-                    &cassandra_protocol::crc::crc32(&dst[payload_start..]).to_le_bytes(),
-                );
+                        // add payload crc
+                        dst.extend_from_slice(&crc32(&dst[payload_start..]).to_le_bytes());
+                    }
+                    Compression::Lz4 => {
+                        let header_start = dst.len();
+
+                        dst.extend_from_slice(&[0, 0, 0, 0, 0, 0, 0, 0]);
+                        let payload_start = dst.len();
+
+                        // TODO we should not be encoding small frames
+                        let (uncompressed_len, compressed_len) =
+                            self.encode_compressed_payload(dst, m, payload_start)?;
+
+                        let mut header =
+                            (compressed_len) as u64 | ((uncompressed_len as u64) << 17);
+
+                        if true {
+                            // TODO if self_contained
+                            header |= 1 << 34;
+                        }
+
+                        let crc = crc24(&header.to_le_bytes()[..5]) as u64;
+
+                        let header = header | (crc << 40);
+
+                        dst[header_start..header_start + 8].copy_from_slice(&header.to_le_bytes());
+
+                        // add payload crc
+                        dst.extend_from_slice(
+                            &cassandra_protocol::crc::crc32(&dst[payload_start..]).to_le_bytes(),
+                        );
+                    }
+                    _ => unimplemented!("Only Lz4 compression is supported for v5"),
+                }
 
                 Ok(())
             }
-            (_, _) => self.encode_envelope(dst, m),
+            (_, _) => {
+                let message_compression = m.codec_state.as_cassandra();
+                self.encode_envelope(dst, m, message_compression)
+            }
         }
     }
 
-    fn encode_envelope(&mut self, dst: &mut BytesMut, m: Message) -> Result<()> {
-        let message_compression = m.codec_state.as_cassandra();
+    fn encode_compressed_payload(
+        &mut self,
+        dst: &mut BytesMut,
+        m: Message,
+        payload_start: usize,
+    ) -> Result<(usize, usize)> {
+        // TODO: always check if cassandra message
+        let bytes = match m.into_encodable(MessageType::Cassandra)? {
+            Encodable::Bytes(bytes) => bytes,
+            Encodable::Frame(frame) => frame
+                .into_cassandra()
+                .unwrap()
+                .encode(Compression::None)
+                .into(),
+        };
+
+        let uncompressed_len = bytes.len();
+        dst.resize(payload_start + get_maximum_output_size(uncompressed_len), 0);
+        let compressed_len = compress_into(&bytes, &mut dst[payload_start..])?;
+        dst.truncate(payload_start + compressed_len);
```

Comment:
Does this have any effect? I would think that the earlier resize has already set it to this size.

Reply:
ooooh right, because `get_maximum_output_size` only provides a maximum, makes sense :+1: